src/csv-upload/service/CsvUploadService.ts

```diff
@@ -0,0 +1,210 @@
+import { Answer } from "$models/Answer";
+import { AnswerRepository } from "$models/Answer/Repository";
+import { Course } from "$models/Course";
+import { CourseRepository } from "$models/Course/Repository";
+import { Question } from "$models/Question";
+import { QuestionRepository } from "$models/Question/Repository";
+import { Semester, SemesterRepository } from "$models/Semester";
+import { Teacher } from "$models/Teacher";
+import { TeacherRepository } from "$models/Teacher/Repository";
+import { TeacherRole } from "$models/TeacherRole";
+import { answersCsvColumns, teachersCsvColumns } from "../csvConstants";
+import { PUBLIC_ANSWERS_WHITELIST } from "./PublicAnswersWhitelist";
+
+export const csvBulkUpsert = async (answers, teachers, year: number, semesterNumber: number) => {
+  const savedSemesters = new Set();
+  const savedCourses = new Set();
+  const savedTeachers = new Set();
+  const savedQuestions = new Set();
+  const savedAnswers = new Set();
+  try {
+    const semester = await getOrCreateSemester(year, semesterNumber);
+    await SemesterRepository.save(semester);
+    savedSemesters.add(semester.uuid);
+    for (const answer of answers) {
+      const course = await getOrCreateCourseByName(
+        answer[answersCsvColumns.EvaluatedConcept],
+        semester
+      );
+      let teacher;
+      if (isEvaluatedElementATeacher(answer[answersCsvColumns.EvaluatedElement])) {
+        teacher = await getOrCreateTeacherByFullName(
+          answer[answersCsvColumns.EvaluatedElement],
+          course
+        );
+      }
+      const question = await getOrCreateQuestion(
+        answer[answersCsvColumns.Question],
+        answer[answersCsvColumns.Block],
+        teacher,
+        course,
+        answer[answersCsvColumns.AnswerValue]
+      );
+      const answerEntity = await getOrCreateAnswer(question, answer[answersCsvColumns.AnswerValue]);
+      answerEntity.count += 1;
+      await CourseRepository.save(course);
+      savedCourses.add(course.uuid);
+      if (teacher) {
+        await TeacherRepository.save(teacher);
+        savedTeachers.add(teacher.uuid);
+      }
+      await QuestionRepository.save(question);
+      savedQuestions.add(question.uuid);
+      await AnswerRepository.save(answerEntity);
+      savedAnswers.add(answerEntity.uuid);
+    }
+    for (const teacherRow of teachers) {
+      const teacher = await getOrCreateTeacherByFullName(teacherRow[teachersCsvColumns.Name], null);
+      teacher.dni = teacherRow[teachersCsvColumns.Dni];
+      await TeacherRepository.save(teacher);
+      savedTeachers.add(teacher.uuid);
+    }
+  } catch (err) {
+    await rollbackCsv(savedSemesters, savedCourses, savedTeachers, savedQuestions, savedAnswers);
```

Comment:
aprovecharia que es idempotente y no haría esto (o si lo hacemos, creo que sería más fácil y performante en vez de guardarnos los `savedFoo`, hacer statements del tipo `DELETE FROM "Questions" (...joins) WHERE "Semester"."year" = 2022 AND "Semester"."semesterNumber" = 1`)

como hablamos en whatsapp, es ideal tener aparte de esto un `deleteSemester` que en cascada borre todos los record asociados a ese semester (para eso es necesario que todo lo que se cree acá pueda ser asociado aun sea indirectamente al semester)

esto es grande, no necesariamente es un comentario sobre el pr sino que puede pasar a tasks en github